design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,137 @@
+# MapLibre Native Code Organisation Design Proposal
+
+This proposal is to define code organisation that allows simple and efficient development and usage of MapLibre Native.
+
+## Motivation
+
+MapLibre Native is currently a monorepo that includes both the core functionality and the SDKs. It's not obvious what is used where andfirst-time contributors may also be repelled a bit by the complexity. This proposal is not necessarily about the splitting of repository (although it will be covered in a section) but in general about code organisation.
+
+## Terminology
+
+- "*renderer*" or "*core*" (will use "*core*" throughout the document) usually refers to the core part of MapLibre Native that is platform-independent to the extent that it builds and runs
+- "*platform*" refers to the operating system and/or a platform that MapLibre runs on, examples are Linux, Windows, macOS, iOS, Android but also Node.js and Qt which provide an universal platform layer that we can use
+- "*SDK*", "*language bindings*" or "*platform bindings*" refer to the interface that is provided to integrate MapLibre Native core to the client applications or projects.
+
+## MapLibre Native Core
+
+MapLibre Native core should provide base rendering functionality and the "brain" of the project. It should implement all the rendering and mapping functionality (i.e. loading of tiles) that is platform-independent. All functionality should be available on all platforms. Platform-specific functionality should be part of the SDKs.
+
+### Organisation
+
+The code needs to be organised in the following way:
+
+- `src`: contains main implementation that is completely platform-independent, also contains all private headers
+- `include`: contains only public headers, that can be split into two categories
+  - functionality of the library
+  - interfaces for platform-specific bindings
+- `platform`: contains minimum set of platform-specific classes for I/O, database access and rendering to make the library functional; any language binding and/or SDK helpers should not go here; examples:
+  - `platform/darwin` or `platform/apple`
+  - `platform/windows`
+  - `platform/qt`
+- `test`: common (headless) tests for all platforms; disabled only in exceptional cases if some features are not applicable due to the nature of a specific platform
+
+The core itself ideally should have no binaries unless they are really platform-independent (e.g. headless rendering).
+
+### Supported platforms
+
+We can split the platforms into two categories:
+
+- **Native platforms / operating systems**
+  - Windows
+  - Linux
+  - Darwin and its derivatives (macOS, iOS, tvOS, ...)
+  - Android
+- **Cross-platform engines**
+  - Node.js
+  - Qt
+  - GLFW
+
+### Build infrastructure
+
+The core should be exclusively built by `CMake`. All currently supported platforms are supported by it. No wrapper script or `Makefile` should be used. Can export `Xcode` projects if needed.
```

Comment:
We compile iOS/macOS under the directory `platform/iOS` and use the existing Xcode project and workspace, so if we fully migrate it to CMake how can we do this just under the command line? Add more compile flags like `MBGL_WITH_QT`?

Reply:
I suspect we keep a Qt and Node.js flag. If you build without flags for a platform it is assumed that you want to build directly for that native platform.

Reply:
I mean add MBGL_WITH_IOS and MBGL_WITH_MACOS to let cmake know which platform should take?

Reply:
And do these platform-depend projects that need to implement by CMake, I think it is not necessary, if we not how let CMake compile it.

Reply:
Not sure I understand the second question.

To answer first, I think it will not be needed as one could just set iOS as the platform (basically cross-compling).